js/ui/appDisplay.js

```diff
@@ -173,13 +173,121 @@ const BaseAppView = new Lang.Class({
         // Nothing by default
     },
 
+    _createItemForId: function(itemId) {
+        if (IconGridLayout.layout.iconIsFolder(itemId))
+            return Shell.DesktopDirInfo.new(itemId);
+
+        return Shell.AppSystem.get_default().lookup_app(itemId);
+    },
+
+    _createItemIcon: function(item) {
+        throw new Error('Not implemented');
+    },
+
+    getLayoutIds: function() {
+        let viewId = this.getViewId();
+        return IconGridLayout.layout.getIcons(viewId).slice();
+    },
+
+    _trimInvisible: function(ids) {
+        let appSystem = Shell.AppSystem.get_default();
+        return ids.filter((id) => {
+            return IconGridLayout.layout.iconIsFolder(id) || appSystem.lookup_app(id) || (id == EOS_APP_CENTER_ID);
+        });
+    },
+
+    _findIconChanges: function() {
+        let oldItemLayout = this._allItems.map(function(icon) { return icon.getId(); });
+        let newItemLayout = this.getLayoutIds();
+        newItemLayout = this._trimInvisible(newItemLayout);
+
+        let movedList = {};
+        let removedList = [];
+        for (let oldItemIdx in oldItemLayout) {
+            let oldItem = oldItemLayout[oldItemIdx];
+            let newItemIdx = newItemLayout.indexOf(oldItem);
+
+            if (oldItemIdx != newItemIdx) {
+                if (newItemIdx < 0)
+                    removedList.push(oldItemIdx);
+                else
+                    movedList[oldItemIdx] = newItemIdx;
+            }
+        }
+
+        return [movedList, removedList];
+    },
+
+    iconsNeedRedraw: function() {
+        // Check if the icons moved around
+        let [movedList, removedList] = this._findIconChanges();
+        let movedLength = Object.keys(movedList).length;
+        if (movedLength > 0 || removedList.length > 0)
+            return true;
+
+        // Create a map from app ids to icon objects
+        let iconTable = {};
+        for (let item of this._allItems)
+            iconTable[item.getId()] = item;
+
+        let layoutIds = this.getLayoutIds();
+
+        // Iterate through all visible icons
+        for (let itemId of layoutIds) {
+            let item = this._createItemForId(itemId);
+            if (!item)
+                continue;
+
+            // The App Center icon cannot be changed or renamed
+            if (item == this._appCenterItem)
+                continue;
+
+            let currentIcon = iconTable[itemId];
+            if (!currentIcon)
+                return true; // This icon is new
+
+            // currentIcon is a ViewIcon (AppIcon or FolderIcon).
+            let oldIconInfo = currentIcon.getIcon();
+
+            // item is either a ShellApp or a  ShellDesktopDirInfo.
+            let newIconInfo = item.get_icon();
+
+            // The icon image changed
+            if (newIconInfo && !newIconInfo.equal(oldIconInfo))
+                return true;
```

Comment:
The old code of this function (https://github.com/endlessm/eos-desktop/blob/master/js/ui/appDisplay.js#L351) also checked whether the name changed. I don't think we still need the bits that use the `customName` field, since that is not set anymore, but I believe we do still need the simple name check from here: https://github.com/endlessm/eos-desktop/blob/master/js/ui/appDisplay.js#L358

Reply:
Good catch. I'll put that extra check in, thanks